### PR TITLE
Database Automatic User Provisioning support for MariaDB

### PR DESCRIPTION
**Pull request by greedy52**
Related
- #27323
- #31902
- #32685

changelog: Database Automatic User Provisioning support for MariaDB

Changes:

- Added support `db.admin_user.default_database`
- Added MariaDB auto-user provisioning stored procedures.
- Enabled MariaDB auto-user provisioning in the engine.

The overall flow is similar to MySQL. Differences are outlined in the comments for `mariadbProcedures`

Next:
- Docs for auto-user provisioning for MySQL/MariaDB

Tested:
- [x] Self-hosted MariaDB 10.2.11 (older than 10.2.11 are not supported)
- [x] Self-hosted MariaDB 10.3.3 (older than 10.3.3 are not supported)
- [x] Self-hosted MariaDB 10.4.0
- [x] Self-hosted MariaDB 10.7
- [x] Self-hosted MariaDB 11.1
- [x] RDS MariaDB instance 10.11
- [x] `Logged in as:       a.very.very.very.very.very.very.very.very.very.very.very.very.long.name@teleport.example.com`

Test Setup Examples:
<details>
<summary>Manual testing example with self-hosted MariaDB </summary>

#### 1. Configure self-hosted MariaDB

Sample docker instance:
https://github.com/greedy52/teleport-database-test-setup/tree/main/mariadb

Log into the database as the default admin then set up Teleport admin user `teleport-admin`:
```
CREATE DATABASE IF NOT EXISTS teleport; -- Default database.
CREATE USER "teleport-admin" REQUIRE SUBJECT "/CN=teleport-admin";
GRANT PROCESS, CREATE USER ON *.* TO 'teleport-admin' ;
GRANT SELECT ON mysql.roles_mapping TO 'teleport-admin' ;
GRANT ALL ON teleport.* TO 'teleport-admin'; -- Allow access to default database.
GRANT UPDATE ON mysql.* TO 'teleport-admin' ; -- Allow SET DEFAULT ROLE FOR.
```
Create a few roles for testing:
```
CREATE ROLE role1 WITH ADMIN 'teleport-admin';
CREATE ROLE role2 WITH ADMIN 'teleport-admin';
CREATE ROLE role3 WITH ADMIN 'teleport-admin';
CREATE DATABASE IF NOT EXISTS test;
GRANT SELECT ON test.* to role1;
```

#### 2. Configure Teleport
Create a Teleport role for auto-user and assign it to your Teleport user, ex:
```yaml
kind: role
version: v6
metadata:
  name: mysql-auto-user
spec:
  options:
    create_db_user: true
  allow:
    db_labels:
      "Owner": "STeve"
      "teleport.dev/db-admin": "teleport-admin"
    db_names:
    - "*"
    db_roles:
    - "role1"
    - "role3"
```
Create a static database in Database Service, ex:
```yaml
db_service:
  enabled: "yes"
  databases:
  - name: "self-hosted-mariadb-auto"
    protocol: "mysql"
    uri: "localhost:3307"
    static_labels:
      Owner: "STeve"
      "teleport.dev/db-admin": "teleport-admin"
    admin_user:
      name: "teleport-admin"
```

#### 3. Connect
- `tsh login`
- `tsh db connect --db-user <teleport-user> --db-name test self-hosted-mariadb-auto`
- `select current_user()`
- `show grants;`
</details>

<details>
<summary>Manual testing example with RDS MariaDB </summary>

#### 1. Configure RDS MariaDB

Setup a MariaDB instance in AWS RDS, and add tags to the RDS instance:
Tags | Value
-- | --
Owner | STeve
teleport.dev/db-admin | teleport-admin

Log into the database as the default admin then set up Teleport admin user `teleport-admin`:
```
CREATE DATABASE IF NOT EXISTS teleport;  -- Default database.
CREATE USER 'teleport-admin' IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS';
GRANT PROCESS, CREATE USER ON *.* TO 'teleport-admin' ;
GRANT SELECT ON mysql.roles_mapping TO 'teleport-admin' ;
GRANT ALL ON teleport.* TO 'teleport-admin'; -- Allow access to default database.
GRANT UPDATE ON mysql.* TO 'teleport-admin' ; -- Allow SET DEFAULT ROLE FOR.
```
Create a few roles for testing:
```
CREATE ROLE role1;
CREATE ROLE role2;
CREATE ROLE role3;
UPDATE mysql.roles_mapping SET User ='teleport-admin' WHERE Admin_option='Y' AND Role='role1';
UPDATE mysql.roles_mapping SET User ='teleport-admin' WHERE Admin_option='Y' AND Role='role2';
UPDATE mysql.roles_mapping SET User ='teleport-admin' WHERE Admin_option='Y' AND Role='role3';
CREATE DATABASE IF NOT EXISTS test;
GRANT SELECT ON test.* to role1;
FLUSH PRIVILEGES;
```
Note that the default admin user in RDS MariaDB does not have permission to do `CREATE ROLE role1 WITH ADMIN teleport-admin`. The workaround is to update `mysql.roles_mapping` manually (and then `FLUSH PRIVILEGES`).

#### 2. Configure Teleport
Create a Teleport role for auto-user and assign it to your Teleport user, ex:
```yaml
kind: role
version: v6
metadata:
  name: mysql-auto-user
spec:
  options:
    create_db_user: true
  allow:
    db_labels:
      "Owner": "STeve"
      "teleport.dev/db-admin": "teleport-admin"
    db_names:
    - "*"
    db_roles:
    - "role1"
    - "role3"
```

Use auto-discovery to register the RD MariaDB database.

#### 3. Connect
- `tsh login`
- `tsh db connect --db-user <teleport-user> --db-name test steve-mariadb`
- `select current_user()`
- `show grants;`
</details>

### CODE REVIEW

**Review by Tener — COMMENTED** [review]
Looks good so far. This is a fairly large change, and the SQL parts are unfamiliar, so I'll do another round on Monday.

File: lib/srv/db/mysql/mariadb_activate_user.sql [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    SET @all_in_one_role := CONCAT("tp-role-", username);
+
+    -- If the user already exists and was provisioned by Teleport, reactivate
+    -- it, otherwise provision a new one.
+    SELECT COUNT(USER) INTO is_auto_user FROM mysql.roles_mapping WHERE Role = 'teleport-auto-user' AND USER = username AND Admin_option = 'N';
+    IF is_auto_user = 1 THEN
+        SELECT COUNT(USER) INTO is_same_user FROM user_attributes WHERE USER = username AND JSON_VALUE(Attributes, "$.user") = @teleport_user;
+        IF is_same_user = 0 THEN
+            SIGNAL SQLSTATE 'TP001' SET MESSAGE_TEXT = 'Teleport username does not match user attributes';
+        END IF;
```
Comment (Tener) [inline lib/srv/db/mysql/mariadb_activate_user.sql:21]: We select from `user_attributes`, but this table is only created below.

File: lib/srv/db/mysql/mariadb_activate_user.sql [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+        -- Set up user attributes.
+        CREATE TABLE IF NOT EXISTS user_attributes(User char(128), Attributes JSON, Primary Key (User));
+        CREATE OR REPLACE VIEW V_user_attributes AS SELECT * FROM user_attributes WHERE SUBSTRING_INDEX(SESSION_USER(),'@',1) = User ;
+        SET @attributes := JSON_EXTRACT(details,"$.attributes");
+        INSERT INTO user_attributes VALUES(username, @attributes) ON DUPLICATE KEY UPDATE Attributes=@attributes;
+        SET @sql := CONCAT_WS(' ', 'GRANT SELECT ON V_user_attributes TO', QUOTE(username));
+        PREPARE stmt FROM @sql;
+        EXECUTE stmt;
+        DEALLOCATE PREPARE stmt;
```
Comment (Tener) [inline lib/srv/db/mysql/mariadb_activate_user.sql:66]: As far as I can see, `V_user_attributes` is not actually used anywhere. We also grant `SELECT` on the view to the user, but I don't see why this is needed.

Overall, it is a bit strange to see table and view creation in the middle of all this; perhaps we should move this all away to the initialization part?

File: lib/srv/db/mysql/autousers.go [hunk 3]
```diff
@@ -469,21 +516,55 @@ var (
...
+	//   while all roles are assigned to this all-in-one role. Then `SET
+	//   DEFAULT ROLE tp-role-<username>` before each session.
+	//
+	// Other MariaDB quirks:
+	// - In order to be able to grant a role, the grantor doing so must have
+	//   permission to do so (see WITH ADMIN in the CREATE ROLE article).
+	//   (Quoted from https://mariadb.com/kb/en/grant/#roles)
+	// - To set a default role for another user one needs to have write access
+	//   to the mysql database.
+	//   (Quoted from https://mariadb.com/kb/en/set-default-role/)
```
Comment (Tener) [inline lib/srv/db/mysql/autousers.go:564]: Nit: I wonder if this comment wouldn't have a better place in the activate procedure itself.

File: lib/srv/db/mysql/mariadb_activate_user.sql (hunk 2 above)
Comment (greedy52) [reply lib/srv/db/mysql/mariadb_activate_user.sql:66]: correct. i was thinking providing a way for users to check their original identity name in their auto-provisioned sessions. not really necessary. will remove

File: lib/srv/db/mysql/autousers.go [hunk 4]
```diff
@@ -319,7 +320,41 @@ func checkMySQLSupportedVersion(serverVersion string) error {
...
+		return nil
+	}
+}
+
+func checkMariaDBSupportedVersion(serverVersion string) error {
+	// serverVersion may look like these:
+	// 5.5.5-10.7.8-MariaDB-1:10.7.8+maria~ubu2004
+	// 5.5.5-10.11.5-MariaDB
+	serverVersion, _, _ = strings.Cut(serverVersion, "-MariaDB")
+	serverVersion = strings.TrimPrefix(serverVersion, "5.5.5-")
```
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/db/mysql/autousers.go]: Couldn't find any official doc on this but added some references.

File: lib/srv/db/mysql/autousers.go (hunk 3 above)
Comment (greedy52) [reply lib/srv/db/mysql/autousers.go]: I was looking for a different place for this but the above go functions are shared for both MySQL/MariaDB. so I decided to keep the comments here.

File: lib/srv/db/mysql/mariadb_activate_user.sql (hunk 1 above)
Comment (Tener) [reply lib/srv/db/mysql/mariadb_activate_user.sql:21]: I see, makes sense, thank you.

File: lib/srv/db/mysql/autousers.go (hunk 3 above)
Comment (Tener) [reply lib/srv/db/mysql/autousers.go]: Fair enough.

**Review by Tener — APPROVED** [review]

**Review by gabrielcorado — APPROVED** [review]

File: api/types/database.go [hunk 5]
```diff
@@ -841,7 +851,7 @@ func (d *DatabaseV3) CheckAndSetDefaults() error {
 
 	// Admin user (for automatic user provisioning) is only supported for
 	// PostgreSQL currently.
```
Comment (smallinsky) [inline api/types/database.go:853]: nit: could you align the comment ?

File: lib/srv/db/mysql/autousers.go [hunk 6]
```diff
@@ -464,25 +515,39 @@ func readDeleteUserResult(res *mysql.Result) string {
 	return string(res.Values[0][0].AsString())
 }
 
+func getCreateProcedureCommand(conn *clientConn, procedureName string) (string, bool) {
+	if conn.isMariaDB() {
+		command, found := mariadbProcedures[procedureName]
+		return command, found
+	}
+	command, found := mysqlProcedures[procedureName]
+	return command, found
```
Comment (smallinsky) [inline lib/srv/db/mysql/autousers.go:524]: nit:
```suggestion
	if conn.isMariaDB() {
		return mariadbProcedures[procedureName]
	}
	return mysqlProcedures[procedureName]
```

**Review by smallinsky — APPROVED** [review]

File: lib/srv/db/mysql/autousers.go (hunk 6 above)
Comment (greedy52) [reply lib/srv/db/mysql/autousers.go:524]: seems go doesn't like this. will keep as it is.

**Pull request merged**